File: tests/runtime/numbers/runner.go
package runner

import (
	"context"
	"fmt"
	"math"

	wrpc "wrpc.io/go"
	"driver/runner/test/numbers/numbers"
)

func Run(ctx context.Context, c wrpc.Invoker) error {
	assertEqual(must(numbers.RoundtripU8(ctx, c, 1)), 1)
	assertEqual(must(numbers.RoundtripU8(ctx, c, 0)), 0)
	assertEqual(must(numbers.RoundtripU8(ctx, c, math.MaxUint8)), math.MaxUint8)

	assertEqual(must(numbers.RoundtripS8(ctx, c, 1)), 1)
	assertEqual(must(numbers.RoundtripS8(ctx, c, math.MinInt8)), math.MinInt8)
	assertEqual(must(numbers.RoundtripS8(ctx, c, math.MaxInt8)), math.MaxInt8)

	assertEqual(must(numbers.RoundtripU16(ctx, c, 1)), 1)
	assertEqual(must(numbers.RoundtripU16(ctx, c, 0)), 0)
	assertEqual(must(numbers.RoundtripU16(ctx, c, math.MaxUint16)), math.MaxUint16)

	assertEqual(must(numbers.RoundtripS16(ctx, c, 1)), 1)
	assertEqual(must(numbers.RoundtripS16(ctx, c, math.MinInt16)), math.MinInt16)
	assertEqual(must(numbers.RoundtripS16(ctx, c, math.MaxInt16)), math.MaxInt16)

	assertEqual(must(numbers.RoundtripU32(ctx, c, 1)), 1)
	assertEqual(must(numbers.RoundtripU32(ctx, c, 0)), 0)
	assertEqual(must(numbers.RoundtripU32(ctx, c, math.MaxUint32)), math.MaxUint32)

	assertEqual(must(numbers.RoundtripS32(ctx, c, 1)), 1)
	assertEqual(must(numbers.RoundtripS32(ctx, c, math.MinInt32)), math.MinInt32)
	assertEqual(must(numbers.RoundtripS32(ctx, c, math.MaxInt32)), math.MaxInt32)

	assertEqual(must(numbers.RoundtripU64(ctx, c, 1)), 1)
	assertEqual(must(numbers.RoundtripU64(ctx, c, 0)), 0)
	assertEqual(must(numbers.RoundtripU64(ctx, c, math.MaxUint64)), math.MaxUint64)

	assertEqual(must(numbers.RoundtripS64(ctx, c, 1)), 1)
	assertEqual(must(numbers.RoundtripS64(ctx, c, math.MinInt64)), math.MinInt64)
	assertEqual(must(numbers.RoundtripS64(ctx, c, math.MaxInt64)), math.MaxInt64)

	assertEqual(must(numbers.RoundtripF32(ctx, c, 1.0)), 1.0)
	assertEqual(must(numbers.RoundtripF32(ctx, c, float32(math.Inf(1)))), float32(math.Inf(1)))
	assertEqual(must(numbers.RoundtripF32(ctx, c, float32(math.Inf(-1)))), float32(math.Inf(-1)))
	assert(math.IsNaN(float64(must(numbers.RoundtripF32(ctx, c, float32(math.NaN()))))))

	assertEqual(must(numbers.RoundtripF64(ctx, c, 1.0)), 1.0)
	assertEqual(must(numbers.RoundtripF64(ctx, c, math.Inf(1))), math.Inf(1))
	assertEqual(must(numbers.RoundtripF64(ctx, c, math.Inf(-1))), math.Inf(-1))
	assert(math.IsNaN(must(numbers.RoundtripF64(ctx, c, math.NaN()))))

	assertEqual(must(numbers.RoundtripChar(ctx, c, 'a')), 'a')
	assertEqual(must(numbers.RoundtripChar(ctx, c, ' ')), ' ')
	assertEqual(must(numbers.RoundtripChar(ctx, c, '🚩')), '🚩')

	must0(numbers.SetScalar(ctx, c, 2))
	assertEqual(must(numbers.GetScalar(ctx, c)), 2)

	must0(numbers.SetScalar(ctx, c, 4))
	assertEqual(must(numbers.GetScalar(ctx, c)), 4)
	return nil
}

func must[T any](v T, err error) T {
	if err != nil {
		panic(err)
	}
	return v
}

func must0(err error) {
	if err != nil {
		panic(err)
	}
}

func assertEqual[T comparable](a T, b T) {
	if a != b {
		panic(fmt.Sprintf("%v not equal to %v", a, b))
	}
}

func assert(v bool) {
	if !v {
		panic("assertion failed")
	}
}

File: tests/runtime/numbers/runner.rs
use crate::client::test::numbers::numbers::*;

pub async fn run(
    wrpc: &impl ::wit_bindgen_wrpc::wrpc_transport::Invoke<Context = ()>,
) -> ::wit_bindgen_wrpc::anyhow::Result<()> {
    assert_eq!(roundtrip_u8(wrpc, (), 1).await?, 1);
    assert_eq!(roundtrip_u8(wrpc, (), u8::MIN).await?, u8::MIN);
    assert_eq!(roundtrip_u8(wrpc, (), u8::MAX).await?, u8::MAX);

    assert_eq!(roundtrip_s8(wrpc, (), 1).await?, 1);
    assert_eq!(roundtrip_s8(wrpc, (), i8::MIN).await?, i8::MIN);
    assert_eq!(roundtrip_s8(wrpc, (), i8::MAX).await?, i8::MAX);

    assert_eq!(roundtrip_u16(wrpc, (), 1).await?, 1);
    assert_eq!(roundtrip_u16(wrpc, (), u16::MIN).await?, u16::MIN);
    assert_eq!(roundtrip_u16(wrpc, (), u16::MAX).await?, u16::MAX);

    assert_eq!(roundtrip_s16(wrpc, (), 1).await?, 1);
    assert_eq!(roundtrip_s16(wrpc, (), i16::MIN).await?, i16::MIN);
    assert_eq!(roundtrip_s16(wrpc, (), i16::MAX).await?, i16::MAX);

    assert_eq!(roundtrip_u32(wrpc, (), 1).await?, 1);
    assert_eq!(roundtrip_u32(wrpc, (), u32::MIN).await?, u32::MIN);
    assert_eq!(roundtrip_u32(wrpc, (), u32::MAX).await?, u32::MAX);

    assert_eq!(roundtrip_s32(wrpc, (), 1).await?, 1);
    assert_eq!(roundtrip_s32(wrpc, (), i32::MIN).await?, i32::MIN);
    assert_eq!(roundtrip_s32(wrpc, (), i32::MAX).await?, i32::MAX);

    assert_eq!(roundtrip_u64(wrpc, (), 1).await?, 1);
    assert_eq!(roundtrip_u64(wrpc, (), u64::MIN).await?, u64::MIN);
    assert_eq!(roundtrip_u64(wrpc, (), u64::MAX).await?, u64::MAX);

    assert_eq!(roundtrip_s64(wrpc, (), 1).await?, 1);
    assert_eq!(roundtrip_s64(wrpc, (), i64::MIN).await?, i64::MIN);
    assert_eq!(roundtrip_s64(wrpc, (), i64::MAX).await?, i64::MAX);

    assert_eq!(roundtrip_f32(wrpc, (), 1.0).await?, 1.0);
    assert_eq!(roundtrip_f32(wrpc, (), f32::INFINITY).await?, f32::INFINITY);
    assert_eq!(
        roundtrip_f32(wrpc, (), f32::NEG_INFINITY).await?,
        f32::NEG_INFINITY
    );
    assert!(roundtrip_f32(wrpc, (), f32::NAN).await?.is_nan());

    assert_eq!(roundtrip_f64(wrpc, (), 1.0).await?, 1.0);
    assert_eq!(roundtrip_f64(wrpc, (), f64::INFINITY).await?, f64::INFINITY);
    assert_eq!(
        roundtrip_f64(wrpc, (), f64::NEG_INFINITY).await?,
        f64::NEG_INFINITY
    );
    assert!(roundtrip_f64(wrpc, (), f64::NAN).await?.is_nan());

    assert_eq!(roundtrip_char(wrpc, (), 'a').await?, 'a');
    assert_eq!(roundtrip_char(wrpc, (), ' ').await?, ' ');
    assert_eq!(roundtrip_char(wrpc, (), '🚩').await?, '🚩');

    set_scalar(wrpc, (), 2).await?;
    assert_eq!(get_scalar(wrpc, ()).await?, 2);
    set_scalar(wrpc, (), 4).await?;
    assert_eq!(get_scalar(wrpc, ()).await?, 4);
    Ok(())
}
